Packet++/header/Asn1Codec.h

```diff
@@ -0,0 +1,462 @@
+		/** The universal tag type for UniversalString */
+		UniversalString = 28,
+		/** The universal tag type for CharacterString */
+		CharacterString = 29,
+		/** The universal tag type for BMPString */
+		BMPString = 30,
+		/** The universal tag type for Date */
+		Date = 31,
+		/** The universal tag type for Time of Day */
+		TimeOfDay = 32,
+		/** The universal tag type for Date-Time */
+		DateTime = 33,
+		/** The universal tag type for Duration */
+		Duration = 34,
+		/** The universal tag type for Object Identifier Internationalized Resource Identifier (IRI) */
+		ObjectIdentifierIRI = 35,
+		/** The universal tag type for Relative Object Identifier Internationalized Resource Identifier (IRI) */
+		RelativeObjectIdentifierIRI = 36,
+		/** A non-applicable value */
+		NotApplicable = 255
+	};
+
+	/**
+	 * @class Asn1Record
+	 * Represents an ASN.1 record, as described in ITU-T Recommendation X.680
+	 */
+	class Asn1Record
+	{
+	public:
+		/**
+		 * A static method to decode a byte array into an Asn1Record
+		 * @param data A byte array to decode
+		 * @param dataLen The byte array length
+		 * @param lazy Use lazy decoding, set to true by default. Lazy decoding entails delaying the decoding
+		 * of the record value until it is accessed
+		 * @return A smart pointer to the decoded ASN.1 record. If the byte stream is not a valid ASN.1 record
+		 * an exception is thrown
+		 */
+		static std::unique_ptr<Asn1Record> decode(const uint8_t* data, size_t dataLen, bool lazy = true);
+
+		/**
+		 * Encode this record and convert it to a byte stream
+		 * @return A vector of bytes representing the record
+		 */
+		std::vector<uint8_t> encode();
+
+		/**
+		 * @return The ASN.1 tag class
+		 */
+		Asn1TagClass getTagClass() const { return m_TagClass; }
+
+		/**
+		 * @return True if it's a constructed record, or false if it's a primitive record
+		 */
+		bool isConstructed() const { return m_IsConstructed; }
+
+		/**
+		 * @return The ASN.1 Universal tag type if the record is of class Universal, otherwise Asn1UniversalTagType#NotApplicable
+		 */
+		Asn1UniversalTagType getUniversalTagType() const;
+
+		/**
+		 * @return The ASN.1 tag type value
+		 */
+		uint8_t getTagType() const { return m_TagType; }
+
+		/**
+		 * @return The length of the record value
+		 */
+		size_t getValueLength() const { return m_ValueLength; }
+
+		/**
+		 * @return The total length of the record
+		 */
+		size_t getTotalLength() const { return m_TotalLength; }
+
+		/**
+		 * A templated method that accepts a class derived from Asn1Record as its template argument and attempts
+		 * to cast the current instance to that type
+		 * @tparam Asn1RecordType The type to cast to
+		 * @return A pointer to the type after casting
+		 */
+		template <class Asn1RecordType>
+		Asn1RecordType* castAs()
+		{
+			auto result = dynamic_cast<Asn1RecordType*>(this);
+			if (result == nullptr)
+			{
+				throw std::runtime_error("Cast failed, instance isn't of the requested type");
```

Comment:
we should use `std::bad_cast` here

Reply:
Fixed in these commits:
https://github.com/seladb/PcapPlusPlus/pull/1353/commits/4b4865f2aa01e3c5ac619f91d7a6d0ff0227934c
https://github.com/seladb/PcapPlusPlus/pull/1353/commits/cb0f1c02a0fef9c012cfc83da7c36df7683f90f7